### PR TITLE
Refactor stop and go test suite

### PR DESCRIPTION
**Pull request by pstjohn**
Refactors the stop and go test suite so that most checks assert that the input/output of an interrupted model run match those trained continuously.

### CODE REVIEW

File: sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py [hunk 1]
```diff
@@ -14,171 +14,239 @@
...
+    Provided methods:
+        - __init__: initializes the callback with the given mode.
+        - get_metadata: abstract method that should be overridden to get metadata from the trainer and pl_module.
+
+    Default behaviors:
+        - in stop mode, metadata is gotten and compared on_validation_epoch_end.
+        - in go mode, metadata is gotten and saved on_train_epoch_start.
+
+    Override these behaviors if necessary.
+    """
```
Comment (pstjohn) [inline sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:213]: I was able to refactor all the remaining callbacks except `TrainValInitConsumedSamplesStopAndGoCallback` to essentially compare states from the interrupted against the continuous training versions.

If we didn't want this callback, or wanted to refactor it, we could probably remove this class

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 2]
```diff
@@ -14,139 +14,129 @@
...
-        model (pl.LightningModule): The PyTorch Lightning model.
-
-    Returns:
-        Any: The global step of the model.
-    """
-    return trainer.global_step
+    return callbacks[mode][callback_type]  # type: ignore
 
 
 class StopAndGoHarness(ABC):
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:61]: Do we need `unittest.TestCase` to make the class methods work?

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 3]
```diff
@@ -160,133 +150,238 @@ def setup_model(
...
+        [
+            testing_callbacks.LearningRateCallback,
+            testing_callbacks.GlobalStepStateCallback,
+            testing_callbacks.ConsumedSamplesCallback,
+            testing_callbacks.OptimizerStateCallback,
+            testing_callbacks.TrainInputCallback,
+            testing_callbacks.TrainOutputCallback,
+            testing_callbacks.TrainLossCallback,
+        ],
+    )
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:335]: Nice

File: sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py [hunk 4]
```diff
@@ -14,171 +14,239 @@
...
+    """Collect training loss samples for comparison."""
+
+    def on_megatron_step_end(
+        self,
+        step: MegatronStep,
+        microbatch_outputs: List[Any],
+        reduced: Optional[Union[torch.Tensor, Dict[str, torch.Tensor]]] = None,
+    ) -> None:
+        """Get consumed samples as metadata."""
+        if step.trainer.training:
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:178]: typo: `validating`

File: sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py [hunk 5]
```diff
@@ -14,171 +14,239 @@
...
 
-    def on_train_batch_start(self, trainer: Trainer, pl_module: LightningModule, batch: Any, batch_idx: int):
-        """PTL callback that raises a StopAndGoException if metadata exists."""
-        pickle_file_path = os.path.join(self.metadata_path, "checkpoints/metadata.pkl")
-        if os.path.exists(pickle_file_path):
-            # Register the signal handler
-            raise StopAndGoException("Terminating early, checkpoint exists.")
-            # kill job afterwards
 
+class BaseInterruptedVsContinuousCallback(Callback, CallbackMethods, io.IOMixin):
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:48]: Nice that we now have this serialized as well.

Comment (sichu2023) [reply sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:48]: Can we add more docstring on `CallbackMethods`? I doubt new users will know what it is.

File: sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py [hunk 6]
```diff
@@ -14,171 +14,239 @@
...
+    def on_train_epoch_start(self, trainer: Trainer, pl_module: LightningModule):  # noqa: D102
+        if self.mode == Mode.RESUME:
+            self.data = self.get_metadata(trainer, pl_module)
+
+    def on_validation_epoch_end(self, trainer: Trainer, pl_module: LightningModule):  # noqa: D102
+        if not trainer.sanity_checking and self.mode == Mode.STOP:
+            self.data = self.get_metadata(trainer, pl_module)
+
+
+class TrainValInitConsumedSamplesStopAndGoCallback(AbstractStopAndGoCallback):
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:243]: @pstjohn Can you highlight why we aren't able to refactor this to use your new base class?

File: sub-packages/bionemo-testing/src/bionemo/testing/torch.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/torch.py:1]: Should we merge this into `utils.py` also?

**Review by sichu2023 — APPROVED** [review]
The only critical comment is the typo.

File: sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py (hunk 4 above)
Comment (pstjohn) [reply sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:178]: good catch! thanks

File: sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py (hunk 6 above)
Comment (pstjohn) [reply sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:243]: done

File: sub-packages/bionemo-testing/src/bionemo/testing/torch.py (hunk 7 above)
Comment (pstjohn) [reply sub-packages/bionemo-testing/src/bionemo/testing/torch.py:1]: Ugh, I hate the name `utils.py` 😄 . I'd prefer not to but I'm flexible. We're not already importing from there in this test suite, so I think it's OK to split them

File: sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py (hunk 5 above)
Comment (pstjohn) [reply sub-packages/bionemo-testing/src/bionemo/testing/testing_callbacks.py:48]: Sure, I added a mention that this is where you can find the available megatron callback functions.

File: sub-packages/bionemo-testing/src/bionemo/testing/torch.py (hunk 7 above)
Comment (sichu2023) [reply sub-packages/bionemo-testing/src/bionemo/testing/torch.py:1]: I am ok with both. `utils.py` sounds too universal.

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 8]
```diff
@@ -14,139 +14,129 @@
...
+
     Stop and go tests act as follows:
-        - setup a clean model for a brief training run, select metrics to track.
-        - interrupt training via the StopAndGoException in the callback InterruptAfterMetadataCallback.
-        - setup a model to be resumed from the checkpoint, with the same metrics.
-        - Restore training and check that metadta matches the stored metrics in the callback CheckpointIntegrityCallback.
-      Useful metrics to check are things like learning rate, global step, validation loss, training loss, and anything
-        else that is important to the training process. If there is an unavailable metrics, a method for fetching the
-        metric should be provided in the bionemo.testing.callbacks module.
+        - setup a clean model for a brief training run, set StopAndGoCallback(s) to track.
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:71]: Rename StopAndGoCallback

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 9]
```diff
@@ -14,139 +14,129 @@
...
+
+    @classmethod
     @abstractmethod
-    def setup_model(
-        self, mode: Literal["stop", "go"]
-    ) -> tuple[pl.LightningModule, pl.LightningDataModule, nl.MegatronOptimizerModule]:
+    def setup_model(cls, mode: Mode) -> tuple[pl.LightningModule, pl.LightningDataModule, nl.MegatronOptimizerModule]:
         """Constructs the model, data, and optimizer for the test harness.
 
         Optionally supports separate code paths for 'stop'/'go', although implementors are
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:142]: Now also support `continuous`.

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py [hunk 10]
```diff
@@ -160,133 +150,238 @@ def setup_model(
...
-    ) -> pl.Trainer:
-        """Constructs the trainer object for the stop and go test.
-
-        This method invokes the get_callbacks method to get the appropriate callbacks for the mode and passes it to the trainer.
+    @classmethod
+    def setup_trainer(
+        cls,
+        mode: Mode,
+    ) -> nl.Trainer:
+        """Setup trainer by passing stop/go callbacks according to mode.
```
Comment (sichu2023) [inline sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:158]: stop, go, and continous.

**Review by jstjohn — APPROVED** [review]

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py (hunk 2 above)
Comment (pstjohn) [reply sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:61]: Right, the base class can't start with `Test` or pytest will try to run it 😄 . I can definitely add a bit to the docstring though

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py (hunk 9 above)
Comment (pstjohn) [reply sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py]: fixed

File: sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py (hunk 8 above)
Comment (pstjohn) [reply sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py:72]: just said `callback`, I think we don't really care what they use

**Pull request merged**